core/cache.hpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2016 Husky Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#pragma once
+
+#include <list>
+#include <unordered_map>
+#include <utility>
+
+#include "boost/optional.hpp"
+
+namespace husky {
+
+template <typename KeyT, typename ValT>
+class ICache {
+   public:
+    explicit ICache(size_t size) { size_ = size; }
+
+    size_t get_max_size() { return size_; }
+
```

Comment:
Can we add some pure virtual functions as the forced interfaces? like `put`, `del`?